x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/terms/index.tsx

```diff
@@ -50,6 +57,79 @@ function isSortableByColumn(layer: IndexPatternLayer, columnId: string) {
   );
 }
 
+function getDisallowedTermsMessage(
+  layer: IndexPatternLayer,
+  columnId: string,
+  indexPattern: IndexPattern
+) {
+  const hasMultipleShifts =
+    uniq(
+      Object.values(layer.columns)
+        .filter((col) => operationDefinitionMap[col.operationType].shiftable)
+        .map((col) => col.timeShift || '')
+    ).length > 1;
+  if (!hasMultipleShifts) {
+    return undefined;
+  }
+  return {
+    message: i18n.translate('xpack.lens.indexPattern.termsWithMultipleShifts', {
+      defaultMessage:
+        "Can't use multiple time shifts in a single layer together with dynamic top values. Either use the same time shift for all metrics or use filters instead of top values.",
+    }),
+    fixAction: {
+      label: i18n.translate('xpack.lens.indexPattern.termsWithMultipleShiftsFixActionLabel', {
+        defaultMessage: 'Pin current top values',
+      }),
+      newState: async (core: CoreStart, frame: FramePublicAPI) => {
+        const fieldName = (layer.columns[columnId] as TermsIndexPatternColumn).sourceField;
+        const response: FieldStatsResponse<string | number> = await core.http.post(
+          `/api/lens/index_stats/${indexPattern.id}/field`,
```

Comment:
I see why you're making this tradeoff now, but as a user I found it confusing that my terms changed size. Maybe we can use `searchSource` instead of the `index_stats` API to fetch the terms, and that would let us provide a size? Or maybe the index_stats API can take a size argument?

Here's a gif of what I experienced where the size increased after pinning:

![time shift more pinned values than shifted](https://user-images.githubusercontent.com/666475/118200476-21de3d80-b423-11eb-8fa7-4da60562c7db.gif)

Reply:
It's hard to make sure there will always be the same terms as matched by the current configuration. I think respecting the current size and allowing to fetch this amount of terms from `index_stats` is a good compromise.